grpc-api/src/main/java/io/deephaven/grpc_api/session/SessionState.java

```diff
@@ -536,17 +595,25 @@ private synchronized void setState(final ExportState state) {
          * @param parent the parent that just resolved; it may have failed
          */
         private void onResolveOne(@Nullable final ExportObject<?> parent) {
+            // am I already cancelled or failed?
+            if (isExportStateTerminal(state)) {
+                return;
+            }
+
             // is this a cascading failure?
-            if (parent != null && isExportStateFinal(parent.state)) {
+            if (parent != null && isExportStateTerminal(parent.state)) {
                 synchronized (this) {
                     errorId = parent.errorId;
+                    ExportState terminalState = ExportState.DEPENDENCY_FAILED;
+
                     if (errorId == null) {
                         final String errorDetails;
                         switch (parent.state) {
                             case RELEASED:
```

Comment:
Should we treat a parent that has been released as `DEPENDENCY_FAILED` for the dependent, or something else?

Reply:
That's a reasonable question. I am thinking of this as a user error - like hey dummy, you released these before you gave me a chance to depend on them. (As well as the cascading cancel -- this failed because a parent (or grandparent) was cancelled.) I'm happy to change to whatever you like. We could just as well do DEPENDENCY_ALREADY_RELEASED to go along with the new ever-accumulating Index of export id's -- this wasn't a reasonable option prior to the index.

Reply:
Let's see what we can do for the user experience of the hapless user who releases their parent dependencies too early, or cancels them, etc.